### ISSUE TITLE
nested dict?

### ISSUE THREAD

**Issue by cjauvin** [honzajavorek/redis-collections#26]
Just to be sure, is this supposed to work:

    d = Dict()
    d['a'] = {}
    d['a']['b'] = 'c'

It seems that it should not, but I've been obtaining conflicting results (still trying to figure out why exactly), and the doc mentions nested Redis Collections, but nothing explicitly about nested regular data structures.

**Comment by russlevy** [comment]
the test above returns <redis_collections.Dict at e4ef356b810e4ede883c37c9f1eaccac {'a': {}}>

where it should return <redis_collections.Dict at e4ef356b810e4ede883c37c9f1eaccac {'a': {'b': 'c'}}>

**Comment by bbayles** [comment]
This is actually a deep problem, and there are unfortunate tradeoffs involved with solving it. I'd be interested in your opinions, @honzajavorek and @russlevy.

A similar situation is discussed in the [documentation for Python's shelve module](https://docs.python.org/3/library/shelve.html). In our case, when `__getitem__` is called it retrieves a pickled value in Redis. When that value is a mutable object like a `dict` or `list`, you can change the contents of what you got back. However, unless you explicitly write them back to Redis, you won't see the changes - `__getitem__` will pull the pickled value out again.

```python
>>> from redis_collections import Dict
# We start with setting a Dict key to a list
>>> redis_dict = Dict([('key', ['list', 'of'])])
# Appending to the list seems to work, but we didn't write anything to Redis!
>>> redis_dict['key'].append('things')
# ...however, the appended item doesn't show up when retrieve the key
>>> redis_dict.items()
0: [('key', ['list', 'of'])]
```

`shelve.Shelf`'s solution to this is the `writeback` parameter. When set to `True` this causes the persistent dictionary to keep a local copy of [what's been retrieved](https://github.com/python/cpython/blob/bfb2841b493ebbb1f4e2045ae64ef43226339cf8/Lib/shelve.py#L109-L117).

There a few downsides to this:
* You have to keep in memory everything that's been fetched from Redis
* You have to write or explicitly [sync](https://docs.python.org/3/library/shelve.html#shelve.Shelf.sync) in order to update Redis, which makes `Dict` less of a substitute for `dict`.
* It's no longer possible to have multiple applications using the same Redis hash and guarantee consistent results. It `'key'` is set to `[1, 2]` in Redis, and application A does `redis_dict['key'].append(3)`, now application B will not get the same result when it does `redis_dict['key']`.

I plan to follow `shelve.Shelf`'s example and to give `Dict` objects a new `cache` object. It will not be on by default, but `Dict(writeback=True)` (or something like that) will activate it.

**Comment by honzajavorek** [comment]
The solution with cache and the alignment with `shelve`'s behaviour sounds good to me. Didn't realise this aspect at first, thanks for explaining the problem! Re-classified this as an improvement.

**Comment by bbayles** [comment]
Workaround implemented in #46!
